File: divide_and_conquer/merge_sort.go
package main

import "fmt"

func merge(l1, l2 []int) []int {
	var newArray []int

	for (len(l1) > 0) && (len(l2) > 0) {
		if l1[0] > l2[0] {
			newArray = append(newArray, l2[0])
			l2 = l2[1:]
		} else {
			newArray = append(newArray, l1[0])
			l1 = l1[1:]
		}
	}

	for len(l1) > 0 {
		newArray = append(newArray, l1[0])
		l1 = l1[1:]
	}

	for len(l2) > 0 {
		newArray = append(newArray, l2[0])
		l2 = l2[1:]
	}

	return newArray
}

func mergeSort(array []int) []int {
	length := len(array)
	if length == 1 {
		return array
	}
	l1 := array[0 : length/2]
	l2 := array[length/2:]

	l1 = mergeSort(l1)
	l2 = mergeSort(l2)

	return merge(l1, l2)
}

func main() {
	a := []int{2, 6, 4, 7, 3, 1, 5}
	x := mergeSort(a[0:])
	fmt.Printf("Final Answer: %v", x)
}

File: divide_and_conquer/merge_sort.py
import math


def merge(a, b):
    combine_list = []
    start = 0
    while len(a) > 0 and len(b) > 0:
        if a[0] > b[0]:
            combine_list.append(b[0])
            b.pop(0)
            start += len(a)
        else:
            combine_list.append(a[0])
            a.pop(0)

    while len(a) > 0:
        combine_list.append(a[0])
        a.pop(0)

    while len(b) > 0:
        combine_list.append(b[0])
        b.pop(0)

    return combine_list, start


def merge_sort(a):
    length = len(a)
    if length == 1:
        return a, 0
    l1 = a[0:math.floor(length/2)]
    l2 = a[math.floor(length/2):]
    l1, x = merge_sort(l1)
    l2, y = merge_sort(l2)
    result, z = merge(l1, l2)

    return result, (x + y + z)


def answer(f):
    values = open(f, 'r')
    test = []
    for line in values:
        tmp = line.split('\n')
        test.append(int(tmp[0]))
    values.close()

    sorted_list, inversions = merge_sort(test)
    print(inversions)


answer('IntegerArray.txt')
